polyply/src/restraints.py

```diff
@@ -12,7 +12,52 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 import networkx as nx
-from .graph_utils import compute_avg_step_length, get_all_predecessors
+from polyply.src.graph_utils import compute_avg_step_length, get_all_predecessors
+
+def upper_bound(avg_step_length, distance, graph_dist, tolerance=0):
+    bound = graph_dist * avg_step_length + distance + tolerance
+    return bound
+
+
+def lower_bound(distance, graph_dist_ref, graph_dist_target, tolerance=0):
+    avg_needed_step_length = distance / graph_dist_target
+    bound = avg_needed_step_length * graph_dist_ref - tolerance
+    return bound
+
+
+def _set_restraint_on_path(graph,
+                           path,
+                           avg_step_length,
+                           distance,
+                           tolerance,
+                           ref_node=None,
+                           target_node=None):
+
+    # graph distances can be computed from the path positions
+    graph_distances_ref = {node: path.index(node) for node in path}
```

Comment:
Expensive! `path.index` costs N for the `N=len(path)`

Reply:
but its cheaper than computing the disjstra path length no?

Reply:
Sure. Consider however, the following:
`{node: idx for (idx, node) in enumerate(path)}`